File: src/skill_xray/llm/session.py
"""One logical-call and input-byte budget shared by both LLM passes."""

import json

from .client import LLMError, LLMResponseError


class LLMBudgetError(LLMError):
    pass


class LLMSession:
    def __init__(self, client, max_calls=25, max_bytes=1 << 20):
        if (type(max_calls) is not int or max_calls < 0
                or type(max_bytes) is not int or max_bytes < 0):
            raise ValueError("LLM budgets must be non-negative integers")
        self.client, self.max_calls, self.max_bytes = client, max_calls, max_bytes
        self.calls = self.input_bytes = self.failures = 0
        self.unavailable = False

    def complete(self, system, user, *, response_schema=None):
        if self.unavailable:
            raise LLMError("session unavailable")
        size = len(system.encode("utf-8")) + len(user.encode("utf-8"))
        if response_schema is not None:
            size += len(json.dumps(response_schema, sort_keys=True, ensure_ascii=False).encode())
        if self.calls >= self.max_calls or self.input_bytes + size > self.max_bytes:
            raise LLMBudgetError("shared LLM budget exhausted")
        self.calls += 1
        self.input_bytes += size
        try:
            structured = getattr(self.client, "complete_structured", None)
            reply = (structured(system, user, response_schema)
                     if response_schema is not None and callable(structured)
                     else self.client.complete(system, user))
            if not isinstance(reply, str) or len(reply.encode("utf-8")) > 16384:
                raise LLMResponseError("response exceeds text budget")
            return reply
        except LLMResponseError:
            self.failures += 1
            raise
        except (LLMError, OSError):
            self.failures += 1
            self.unavailable = True
            raise LLMError("session transport failure") from None
        except Exception:
            self.failures += 1
            raise

    def usage(self):
        identity = {}
        for name, default in (("provider", "custom"), ("model", "unknown")):
            try:
                value = getattr(getattr(self.client, "cfg", None), name, default)
            except Exception:
                value = None
            identity[name] = (value if type(value) is str and 0 < len(value) <= 200
                              and value == value.strip() and value.isprintable() else "unknown")
        return {"calls": self.calls, "input_bytes": self.input_bytes,
                **identity,
                "failures": self.failures, "unavailable": self.unavailable,
                "max_calls": self.max_calls, "max_input_bytes": self.max_bytes,
                "unit": "logical-completions; transport retries remain separately bounded"}

File: internal/llm/session.go
package llm

import (
	"errors"
	"net"
	"unicode/utf8"

	"github.com/traceforce/skill-xray/internal/pytext"
)

// Session is LLMSession: one logical-call and input-byte budget shared by the review and the
// advisory pass. A transport failure marks it unavailable for both.
type Session struct {
	Client                                          Completer
	MaxCalls, MaxBytes, Calls, InputBytes, Failures int
	Unavailable                                     bool
}

// NewSession is LLMSession(client, max_calls, max_bytes); the budgets must be non-negative.
func NewSession(c Completer, maxCalls, maxBytes int) (*Session, error) {
	if maxCalls < 0 || maxBytes < 0 {
		return nil, errors.New("LLM budgets must be non-negative integers")
	}
	return &Session{Client: c, MaxCalls: maxCalls, MaxBytes: maxBytes}, nil
}

// structuredCompleter and identified are the optional client surface (Python getattr duck typing).
type structuredCompleter interface {
	CompleteStructured(system, user string, schema any) (string, error)
}
type identified interface {
	Identity() (provider, model string)
}

// Complete satisfies Completer, so Adjudicate runs through the shared budget.
func (s *Session) Complete(system, user string) (string, error) { return s.complete(system, user, nil) }

// complete is LLMSession.complete: budget first, then the call, then the reply bound. A Response
// error passes through; a transport (or OS-level) failure disables the session and is replaced by
// a fixed message so a credential in the original text never propagates.
func (s *Session) complete(system, user string, schema any) (string, error) {
	if s.Unavailable {
		return "", &Error{Transport, "session unavailable"}
	}
	size := len(system) + len(user)
	if schema != nil {
		// ponytail: Python dumps with ensure_ascii=False; identical for the ASCII response schema.
		size += len(pytext.Dumps(schema, 0))
	}
	if s.Calls >= s.MaxCalls || s.InputBytes+size > s.MaxBytes {
		return "", &Error{Budget, "shared LLM budget exhausted"}
	}
	s.Calls++
	s.InputBytes += size
	var reply string
	var err error
	if sc, ok := s.Client.(structuredCompleter); ok && schema != nil {
		reply, err = sc.CompleteStructured(system, user, schema)
	} else {
		reply, err = s.Client.Complete(system, user)
	}
	if err == nil && len(reply) > 16384 {
		err = &Error{Response, "response exceeds text budget"}
	}
	if err == nil {
		return reply, nil
	}
	s.Failures++
	var e *Error
	var netErr net.Error // ponytail: OSError is net.Error here; a custom client's file errors classify as "error"
	switch {
	case errors.As(err, &e) && e.Kind == Response:
		return "", err
	case errors.As(err, &e) || errors.As(err, &netErr):
		s.Unavailable = true
		return "", &Error{Transport, "session transport failure"}
	}
	return "", err
}

// Usage is LLMSession.usage(): the enrichment.llm_usage record. The client identity is kept only
// when it is a bounded, stripped, printable string; otherwise "unknown".
func (s *Session) Usage() map[string]any {
	provider, model := "custom", "unknown"
	if id, ok := s.Client.(identified); ok {
		provider, model = id.Identity()
	}
	bounded := func(v string) string {
		if n := utf8.RuneCountInString(v); 0 < n && n <= 200 && v == pytext.Strip(v) && pytext.IsPrintable(v) {
			return v
		}
		return "unknown"
	}
	return map[string]any{"calls": s.Calls, "input_bytes": s.InputBytes, "provider": bounded(provider),
		"model": bounded(model), "failures": s.Failures, "unavailable": s.Unavailable, "max_calls": s.MaxCalls,
		"max_input_bytes": s.MaxBytes, "unit": "logical-completions; transport retries remain separately bounded"}
}
